### DISCUSSION TITLE
Snippet icon looks like a broken/missing icon

### DISCUSSION THREAD

**Discussion by memeplex** [company-mode/company-mode#1233] [General]
The first time I opened a company completion popup after installing doom emacs I got this:

![image](https://user-images.githubusercontent.com/2845433/135768294-d2044542-c819-44e9-92e4-91d5732533d7.png)

So I reported it in https://github.com/hlissner/doom-emacs/issues/5563 just to realize it isn't a broken icon but a real SVG one.

Would you agree with me in that the little black box looks like if the icon were missing? Don't you think a different default would be more sensible?

Best regards,
Carlos

**Comment by dgutov** [comment]
I can understand where you're coming from, but this is an icon from the VS Code icon set: https://code.visualstudio.com/docs/editor/intellisense#_types-of-completions I imagine they do a lot of testing so it must be okay for a lot of people.

That said, replacing it is not out of the question, but the replacement needs to be a similarly high quality SVG icon, and it needs to fit the rest of the set.

**Comment by memeplex** [comment]
Ok, I see, maybe it was because the first time I only see that icon (all available completions were snippets) that I thought something had to be broken.

**Comment by ville-h** [comment]
I'll pile on and say I thought it was broken/placeholder for a glyph my font doesn't have and had put as a "thing to fix" on my list. I was reading the discussions-secion trying to find a fix for something else and ran into this thread by accident.

If it's for snippets then would unicode {black,white} scissors be alright: ✂, ✄?

**Reply by dgutov** [reply]
Thanks for your feedback, but I'd rather not change the snippet icon alone to something not in VSCode's set.

But note that you can still change it for yourself by changing `company-vscode-icons-mapping` and setting the cdr of `snippet`'s cons to be an absolute file name pointing to a different icon.

**Reply by ville-h** [reply]
The `company-vscode-icons-mapping` does not appear to be documented. Does the value in: `(name . "value")` have to be an svg filename or can it be something else? If those have to be svg filenames then is there a "higher-level" variable that can be used to control that most things use the provided icons but some may be set to use text such as the unicode scissors my font provides?

Also seems slightly odd to name the variable after a specific product if it's meant as a general customization point for users to include things which aren't tied to the named product. The tool called "lldb-vscode" comes to mind as an example from elsewhere.

**Reply by dgutov** [reply]
Yes, it was not made to be customizable because it also depends on the files structure inside `company-icons-root`, in particular, the presence of two directories in it: `vscode-light` and `vscode-dark`.

If you need to use something other than `svg` anyway, it'll be easier to define your own `company-format-margin-function`: it should `(company-call-backend 'kind candidate)` and, depending on the return value, either return a string containing your unicode character, or delegate to `company-detect-icons-margin` (or a specific margin function of your choice).
